### PR TITLE
PLANNER-2483 Only build KieBase once per constraint factory

### PR DESCRIPTION
**Pull request by triceo**
Depends on DROOLS-6510 to be merged and Kogito upgraded.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java [hunk 1]
```diff
@@ -42,22 +37,6 @@ protected AbstractConstraintStreamScoreDirectorFactory(SolutionDescriptor<Soluti
...
-                    + ")'s defineConstraints() must not return null.\n"
-                    + "Maybe return an empty array instead if there are no constraints.");
-        }
-        if (Arrays.stream(constraints).anyMatch(Objects::isNull)) {
-            throw new IllegalStateException("The constraintProvider class (" + constraintProvider.getClass()
-                    + ")'s defineConstraints() must not contain an element that is null.\n"
-                    + "Maybe don't include any null elements in the " + Constraint.class.getSimpleName() + " array.");
-        }
-        return constraints;
-    }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java:59]: Moved to `...ConstraintFactory`, where it makes logical sense.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java [hunk 2]
```diff
@@ -45,17 +85,103 @@ public DroolsConstraintStreamScoreDirectorFactory(SolutionDescriptor<Solution_>
         return new DroolsConstraintStreamScoreDirector<>(this, lookUpEnabled, constraintMatchEnabledPreference);
     }
 
-    public DroolsConstraintSessionFactory.SessionDescriptor<Score_>
-            newConstraintStreamingSession(boolean constraintMatchEnabled, Solution_ workingSolution) {
-        return constraintSessionFactory.buildSession(constraintMatchEnabled, workingSolution);
+    public static <Solution_> KieBaseDescriptor<Solution_> buildKieBase(SolutionDescriptor<Solution_> solutionDescriptor,
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java:88]: This is prepared to be used by Quarkus from the outside of the class hierarchy.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintFactory.java [hunk 3]
```diff
@@ -55,30 +50,6 @@ public BavetConstraintFactory(SolutionDescriptor<Solution_> solutionDescriptor)
...
-            if (!added) {
-                throw new IllegalStateException(
-                        "There are 2 constraints with the same constraintName (" + constraint.getConstraintName()
-                                + ") in the same constraintPackage (" + constraint.getConstraintPackage() + ").");
-            }
-            BavetConstraint<Solution_> bavetConstraint = (BavetConstraint) constraint;
-            bavetConstraintList.add(bavetConstraint);
-        }
-        return new BavetConstraintSessionFactory<>(solutionDescriptor, bavetConstraintList);
-    }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintFactory.java:80]: This logic is moved to constraint factory.

**Comment by triceo** [comment]
The GHA failures are caused by the fact that new Drools had only just been merged to Kogito. They will be gone tomorrow, when new Kogito SNAPSHOTs are available.

**Review by rsynek — APPROVED** [review]

**Comment by triceo** [comment]
Apps failure unrelated, merging.

**Pull request merged**